storage/inmem/txn.go

```diff
@@ -0,0 +1,410 @@
+		if len(curr.path) > 0 {
+			apply(txn.db.data, curr.path, curr.remove, curr.value)
+		} else {
+			txn.db.data = curr.value.(map[string]interface{})
+		}
+		result.SetDataChanged()
+	}
+	for id, bs := range txn.policies {
+		if bs == nil {
+			delete(txn.db.policies, id)
+		} else {
+			txn.db.policies[id] = bs
+		}
+		result.SetPolicyChanged()
+	}
+	return result
+}
+
+func (txn *transaction) read(path storage.Path) (interface{}, error) {
+
+	if !txn.write {
+		return ptr(txn.db.data, path)
+	}
+
+	merge := []*snapshot{}
+
+	for curr := txn.log; curr != nil; curr = curr.next {
+		if path.HasPrefix(curr.path) {
+			if curr.remove {
+				return nil, notFoundError(path)
+			}
+			return ptr(curr.value, path[len(curr.path):])
+		}
+		if curr.path.HasPrefix(path) {
+			merge = append(merge, curr)
+		}
+	}
+
+	data, err := ptr(txn.db.data, path)
+
+	if err != nil {
+		if !storage.IsNotFound(err) || len(merge) == 0 {
+			return nil, err
+		}
+	}
+
+	if len(merge) == 0 {
+		return data, nil
+	}
+
+	cpy := deepCopy(data)
+
+	for _, s := range merge {
+		suffix := s.path[len(path):]
+		apply(cpy, suffix, s.remove, s.value)
+	}
+
+	return cpy, nil
+}
+
+func (txn *transaction) listPolicies() []string {
+	var ids []string
+	for id := range txn.db.policies {
+		if bs, ok := txn.policies[id]; !ok || bs != nil {
+			ids = append(ids, id)
+		}
+	}
+	for id, bs := range txn.policies {
+		if bs != nil {
+			if _, ok := txn.db.policies[id]; !ok {
+				ids = append(ids, id)
+			}
+		}
+	}
+	return ids
+}
+
+func (txn *transaction) getPolicy(id string) ([]byte, error) {
+	if exist, ok := txn.policies[id]; ok {
+		if exist != nil {
+			return exist, nil
+		}
+		return nil, notFoundErrorf("policy id %q", id)
+	}
+	if exist, ok := txn.db.policies[id]; ok {
+		return exist, nil
+	}
+	return nil, notFoundErrorf("policy id %q", id)
+}
+
+func (txn *transaction) setPolicy(id string, bs []byte) error {
+	if !txn.write {
+		return &storage.Error{
+			Code:    storage.InvalidTransactionErr,
+			Message: "policy write during read transaction",
+		}
+	}
+	txn.policies[id] = bs
+	return nil
+}
+
+func apply(data interface{}, path storage.Path, remove bool, value interface{}) {
+	parent := getUnsafe(data, path[:len(path)-1])
+	key := path[len(path)-1]
+	if remove {
+		obj := parent.(map[string]interface{})
+		delete(obj, key)
+		return
+	}
+	switch parent := parent.(type) {
+	case map[string]interface{}:
+		parent[key] = value
+	case []interface{}:
+		parent[indexUnsafe(key)] = value
+	}
+}
+
+func deepCopy(val interface{}) interface{} {
+	switch val := val.(type) {
+	case []interface{}:
+		cpy := make([]interface{}, len(val))
+		for i := range cpy {
+			cpy[i] = deepCopy(val[i])
+		}
+		return cpy
+	case map[string]interface{}:
+		cpy := make(map[string]interface{}, len(val))
+		for k := range val {
+			cpy[k] = deepCopy(val[k])
+		}
+		return cpy
+	default:
+		return val
+	}
+}
+
+func getUnsafe(data interface{}, path storage.Path) interface{} {
```

Comment:
not sure if the unsafe here captures the right intent. at least I was assuming golang unsafe pkg right away...